src/mscorlib/shared/System/Globalization/CompareInfo.cs

```diff
@@ -1217,6 +1219,54 @@ public override int GetHashCode()
             return (this.Name.GetHashCode());
         }
 
+        internal static unsafe int GetIgnoreCaseHash(string source)
+        {
+            Debug.Assert(source != null, "source must not be null");
+
+            // Do not allocate on the stack if string is empty
+            if (source.Length == 0)
+            {
+                return source.GetHashCode();
+            }
+
+            char[] borrowedArr = null;
+            Span<char> span = source.Length <= 255 ?
+                stackalloc char[255] :
+                (borrowedArr = ArrayPool<char>.Shared.Rent(source.Length));
+
+            char c;
+            // We assume that str contains only ASCII characters until
+            // we hit a non-ASCII character to optimize the common case.
+            for (int i = 0; i < source.Length; i++)
+            {
+                c = source[i];
+
+                if (c >= 0x80)
+                {
+                    source.AsSpan().ToUpper(span, CultureInfo.InvariantCulture);
```

Comment:
>source.AsSpan().ToUpper(span, CultureInfo.InvariantCulture); [](start = 20, length = 60)

if you want to optimize more, you may upper case the range starting from i and not the whole string

Reply:
This optimization is done by ToUpperInvariant already (on some paths at least).

Should we just call `source.AsSpan().ToUpperInvariant` here instead of trying to replicate all the optimizations from it here? If there are optimizations missing in ToUpperInvariant, we should add them there - not trying to reinvent them here.

Reply:
thanks